_posts/2021-06-01-clean-control-flow-in-elixir-with-pattern-matching-and-immutability.md

```diff
@@ -0,0 +1,133 @@
+---
+author: Cristine Guadelupe
+author_link: https://github.com/cristineguadelupe
+categories: general
+date: 2021-01-06
+layout: post
+title: "Clean Control Flow in Elixir with Pattern Matching and Immutability"
+excerpt: >
+  Learn how to you use pattern matching instead of guard clauses to implement a really clean control flow in Elixir.
+---
+One of the features that fascinate me most about Elixir is pattern matching. I always wonder if there is a way to solve what I need using it, and I love exploring it. When you combine the beauty of pattern matching with the power of immutability, some things almost seem magical. But they are not!
+It is not my focus to cover everything about pattern matching and immutability on this post. My goal here is to show you how we can use pattern matching instead of guard clauses to implement a really clean control flow in Elixir.
+
+Let's say you're building the game of battleship and you need to build some logic that says the same player can't attack twice in a row. One way to solve the problem is to register the identifier of the last player that performed an attack.
+With this information we now have two possibilities, if the player who is trying to make a new attack is the same as the last one that was registered as the last shooter, we will just ignore the attack. Otherwise, if the attack is coming from another player, we will compute the attack.
+
+Depending on your elixir knowledge, you might think of some conditional as the first solution, something like:
+
+```elixir 
+def maybe_attack(player, last_shooter) do
+    if player != last_shooter do
+        player
+        |> make_an_attack()
+        |> set_as_last_shooter()
+    else
+      ignore()
+    end
+end
+```
+
+Or even pattern matching with guard clause
+
+```elixir
+def maybe_attack(player, last_shooter) when player == last_shooter do
+    ignore()
+end
+
+def maybe_attack(player, last_shooter) do
+    player
+    |> make_an_attack()
+    |> set_as_last_shooter()
+end
+```
+
+But it is possible to combine the pattern matching we already used in the guard clause solution with the power of immutability to come up with an even more alchemistic solution!
+
+```elixir
+def maybe_attack(last_shooter, last_shooter) do
+    ignore()
+end
+
+def maybe_attack(player, last_shooter) do
+    player
+    |> make_an_attack()
+    |> set_as_last_shooter()
+end
+```
+
+Wait a second, what have we done here?
+
+In the first `maybe_attack` function we are calling our player (the first argument) as last_shooter!
```

Comment:
I would rephrase this slightly to read something like "We define the first version of the `maybe_attack` function to take in a first _and_ second argument named `last_shooter`. This means the function will only match if the player provided as a first argument matches the player provided as a second argument". Pretty much same thing you are saying here, but just makes the language a little more clear 😄